packages/headless/src/features/folding/folding-slice.ts

```diff
@@ -0,0 +1,125 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {Result} from '../../api/search/search/result';
+import {isArray, removeDuplicates} from '../../utils/utils';
+import {executeSearch, fetchMoreResults} from '../search/search-actions';
+import {registerFolding} from './folding-actions';
+import {
+  FoldedResult,
+  FoldingFields,
+  getFoldingInitialState,
+} from './folding-state';
+
+interface ResultWithFolding extends Result {
+  parentResult?: ResultWithFolding;
+  childResults?: ResultWithFolding[];
+}
+
+function getRawValue(result: ResultWithFolding, field: string) {
+  const rawValue = result.raw[field] as string | [string] | undefined;
+  if (isArray(rawValue)) {
+    return rawValue[0];
+  }
+  return rawValue;
+}
+
+function areDefinedAndEqual<T>(
+  value1: T | undefined,
+  value2: T | undefined
+): boolean {
+  return (value1 || value2) !== undefined && value1 === value2;
+}
+
+function flattenResultTree(result: ResultWithFolding): ResultWithFolding[] {
```

Comment:
Since the original order of `parentResult` and `childResults` is wrong (a child and a sibling can both be listed in the same `childResults` array), the ultimate source of truth is the `parent` and `child` raw fields.

I felt like it would be simpler and less prone to regression to flatten the entire result tree and then re-build it in the folded order.